components/AppBar/src/MDCAppBarNavigationController.m

```diff
@@ -0,0 +1,193 @@
+  info.appBar = appBar;
+  info.trackingScrollView = trackingScrollView;
+  [self setInfo:info forViewController:viewController];
+
+  // Ensures that the view controller's top layout guide / additional safe area insets are adjusted
+  // to take into consideration the flexible header's height.
+  appBar.headerViewController.topLayoutGuideViewController = viewController;
+
+  // Ensures that our App Bar's top layout guide reflects the current view controller hierarchy.
+  // Most notably, this ensures we support iPad popovers and extensions.
+  appBar.headerViewController.inferTopSafeAreaInsetFromViewController = YES;
+
+  // We want our flexible header to calculate the safe area insets dynamically, rather than assume
+  // we've pre-calculated them.
+  appBar.headerViewController.headerView.minMaxHeightIncludesSafeArea = NO;
+
+  // This is the magic that allows us to avoid having to explicitly forward any scroll view events
+  // to the flexible header. Enabling this means we cannot enable the shiftBehavior on the
+  // flexible header. In those cases the client is expected to create their own App Bar.
+  appBar.headerViewController.headerView.observesTrackingScrollViewScrollEvents = YES;
+
+  appBar.headerViewController.headerView.trackingScrollView = trackingScrollView;
+
+  if ([self.delegate respondsToSelector:
+       @selector(appBarNavigationController:willAddAppBar:asChildOfViewController:)]) {
+    [self.delegate appBarNavigationController:self
+                                willAddAppBar:appBar
+                      asChildOfViewController:viewController];
+  }
+
+  [viewController addChildViewController:appBar.headerViewController];
+  [appBar addSubviewsToParent];
+}
+
+- (BOOL)viewControllerHasFlexibleHeader:(UIViewController *)viewController {
+  // Searching the child view controllers covers both the contained case and the injected-as-a-child
+  // case. MDCFlexibleHeaderViewController can never be a top-level view controller.
+  for (UIViewController *childViewController in viewController.childViewControllers) {
+    if ([childViewController isKindOfClass:[MDCFlexibleHeaderViewController class]]) {
+      return YES;
+    }
+  }
+  return NO;
+}
+
+- (UIScrollView *)findFirstInstanceOfUIScrollViewInView:(UIView *)view {
+  if ([view isKindOfClass:[UIScrollView class]]) {
+    return (UIScrollView *)view;
+  }
+  for (UIView *subview in view.subviews) {
+    UIScrollView *scrollView = [self findFirstInstanceOfUIScrollViewInView:subview];
+    if (scrollView != nil) {
+      return scrollView;
+    }
+  }
+  return nil;
+}
+
+- (MDCAppBarNavigationControllerInfo *)infoForViewController:(UIViewController *)viewController {
+  return objc_getAssociatedObject(viewController, _cmd);
+}
+
+- (void)setInfo:(MDCAppBarNavigationControllerInfo *)info
+    forViewController:(UIViewController *)viewController {
+  objc_setAssociatedObject(viewController,
```

Comment:
These are terrifying - do they go away in dealloc?

Reply:
They do - as defined by the associated memory type below.